exercises/practice/rest-api/.meta/Example.roc

```diff
@@ -0,0 +1,214 @@
+module [get, post]
+
+import json.Json
+
+User : {
+    name : Str,
+    owes : Dict Str F64,
+    owedBy : Dict Str F64,
+    balance : F64,
+}
+
+Database : { users : List User }
+
+Loan : { lender : Str, borrower : Str, amount : F64 }
+
+## Handle GET requests
+get : Database, { url : Str, payload ? Str } -> Result Str [Http404 Str, Http422 Str]
+get = \database, { url, payload ? "" } ->
+    when url is
+        "/users" ->
+            database
+            |> getUsers payload
+            |> Result.mapErr \err ->
+                when err is
+                    InvalidJson -> Http422 payload
+
+        badUrl -> Err (Http404 badUrl)
+
+## Compare two strings, first by their UTF8 representations, then by length:
+## "" < "ABC" < "abc" < "abcdef"
+## This is used to sort the users in the JSON outputs
+compareStrings : Str, Str -> [LT, EQ, GT]
+compareStrings = \string1, string2 ->
+    b1 = string1 |> Str.toUtf8
+    b2 = string2 |> Str.toUtf8
+    result =
+        List.map2 b1 b2 \c1, c2 -> Num.compare c1 c2
+        |> List.walkTry (Ok EQ) \_state, cmp ->
+            when cmp is
+                EQ -> Ok EQ
+                res -> Err res
+    when result is
+        Ok _cmp -> Num.compare (List.len b1) (List.len b2)
+        Err res -> res
+
+## Convert the owes and owedBy dictionaries to JSON
+## Note: this will be simpler once Roc supports Encoding & Decoding Dict
+oweDictToJson : Dict Str F64 -> Str
+oweDictToJson = \dict ->
+    dictContent =
+        dict
+        |> Dict.toList
+        |> List.sortWith \(name1, _amount1), (name2, _amount2) -> compareStrings name1 name2
+        |> List.map \(name, amount) -> "$(name |> stringToJson): $(amount |> Num.toStr)"
+        |> Str.joinWith ","
+    "{$(dictContent)}"
+
+stringToJson : Str -> Str
+stringToJson = \string ->
+    result =
+        Encode.toBytes string Json.utf8
+        |> Str.fromUtf8
+    when result is
+        Ok json -> json
+        Err (BadUtf8 _ _) -> crash "Unreachable: encoding a string to JSON should never fail"
+
+## Convert a user to a JSON representation
+## Note: this will be simpler once Roc supports Encoding & Decoding Dict
+userToJson : User -> Str
+userToJson = \user ->
+    """
+    {
+        \"balance\" : $(user.balance |> Num.toStr),
+        \"name\" : $(user.name |> stringToJson),
+        \"owed_by\" : $(user.owedBy |> oweDictToJson),
+        \"owes\" : $(user.owes |> oweDictToJson)
```

Comment:
I don't think these `"` need to be escaped

Reply:
Good point, I'll fix that.